prometheus_hardware_exporter/core.py

```diff
@@ -111,21 +127,28 @@ def collect(self) -> Iterable[Metric]:
         Yields:
             metrics: the internal metrics
         """
-        payloads = self.fetch()
-        self.init_default_datastore(payloads)
-        processed_payloads = self.process(payloads, self._datastore)
-
-        # unpacked and create metrics
-        for payload in processed_payloads:
-            spec = self._specs[payload.name]
-            # We have to ignore the type checking here, since the subclass of
-            # any metric family from prometheus client adds new attributes and
-            # methods.
-            metric = spec.metric_class(  # type: ignore[call-arg]
-                name=spec.name, labels=spec.labels, documentation=spec.documentation
-            )
-            metric.add_metric(  # type: ignore[attr-defined]
-                labels=payload.labels, value=payload.value
-            )
-            yield metric
-            self._datastore[payload.uuid] = payload
+        # The general exception hanlder will try to make sure the single
+        # collector's bug will only change the metrics output to failed_metrics
+        # and also make sure other collectors are still working.
+        try:
+            payloads = self.fetch()
+            self.init_default_datastore(payloads)
+            processed_payloads = self.process(payloads, self._datastore)
+
+            # unpacked and create metrics
+            for payload in processed_payloads:
+                spec = self._specs[payload.name]
+                # We have to ignore the type checking here, since the subclass of
+                # any metric family from prometheus client adds new attributes and
+                # methods.
+                metric = spec.metric_class(  # type: ignore[call-arg]
+                    name=spec.name, labels=spec.labels, documentation=spec.documentation
+                )
+                metric.add_metric(  # type: ignore[attr-defined]
+                    labels=payload.labels, value=payload.value
+                )
+                yield metric
+                self._datastore[payload.uuid] = payload
+        except Exception as e:  # pylint: disable=W0718
+            logger.error(e)
+            yield from self.failed_metrics
```

Comment:
Would it not be better if failed metrics contains some information about why collection failed? Maybe via label?

Reply:
It's for unexpected error. So it's weird to have a label of it.

Reply:
It will mean that we will have more exception not only the generic one, if it's possible. e.g.
```python
except ExceptionA as exc:
    logger.exception(exc)
    yield from self.failed_metrics(label="ExceptionA")
except ExceptionB as exc:
    logger.exception(exc)
    yield from self.failed_metrics(label="ExceptionB")
except Exception as exc:
    logger.exception(exc)
    yield from self.failed_metrics(label="Unknown")
```

Reply:
If that's the case, it should be handle inside each controller from my view.

Reply:
I'm not an expert here, but I do not understand what is controller here and how it could handle those exceptions.

Reply:
The purpose is that we try to catch those unexpected error and try not to make exporter whole broken. It's kind of defensive purpose.

So if we know what exactly the error is, we should handle that inside collector.

Reply:
Ok, that make sense to me. Thanks

Reply:
And I think you meant collector by "each controller from my view."  as I understand it from https://github.com/canonical/hardware-observer-operator/pull/116.